src/Weeder/Main.hs

```diff
@@ -105,7 +105,15 @@ main = do
 -- This will recursively find all files with the given extension in the given directories, perform
 -- analysis, and report all unused definitions according to the 'Config'.
 mainWithConfig :: String -> [FilePath] -> Bool -> Config -> IO ()
-mainWithConfig hieExt hieDirectories requireHsFiles Config{ rootPatterns, typeClassRoots } = do
+mainWithConfig hieExt hieDirectories requireHsFiles weederConfig = 
+  mainWithConfig' hieExt hieDirectories requireHsFiles weederConfig 
+    >>= \(success, _) -> unless success exitFailure
+
+
+-- | Alternative mainWithConfig that returns the analysis and whether a zero exit
+-- code should be returned.
+mainWithConfig' :: String -> [FilePath] -> Bool -> Config -> IO (Bool, Analysis)
+mainWithConfig' hieExt hieDirectories requireHsFiles Config{ rootPatterns, typeClassRoots } = do
```

Comment:
I'm not very big on `'` naming, and we now end up with three `main*` functions. This seems a bit confusing. How about we go with just having `mainWithConfig` return `(ExitCode, Analysis)`. Then `main` can be responsible for calling `exitWith`, and your test suite can check `ExitCode` too.

So to recap:

* `main` calls `mainWithConfig` and then calls `exitWith` with the `ExitCode` that `mainWithConfig` returns
* `mainWithConfig` reurns `(ExitCode, Analysis)`. The `ExitCode` is formatted based on `null warnings`, as before.
* We drop the new `mainWithConfig'`.